Feature/3.0

boralyl opened the pull request.

Commit:
Adding python types to describe and validate request parameters.

boralyl committed:
Added a few tests for types.

Added missing exception class.  Updated setup.cfg w/ pytest values

boralyl committed:
Added tests for all types.

boralyl committed:
Added types documentation

Commit:
Don't build python2.7

Commit:
Added missing JsonSchema documentation

Commit:
Fixed some typos

boralyl committed:
Merge pull request #40 from upsight/feature/add-types

Feature/add types

Commit:
Added format attribute logic to String type

Commit:
Updating comment

Commit:
Renamed jsonschematype and newtype to match our conventions.

Commit:
Adding new code for creating routes using python types.

Commit:
Added new router and handle http logic.

Commit:
Moved the new example app into it's own file.

Modifying the existing app caused test failures and I'm not ready to
address documentation just yet.

Commit:
Added tests for the handle_http_v3

boralyl committed:
Merge pull request #41 from upsight/feature/string-type-format

Feature/string type format

boralyl committed:
Merge pull request #44 from upsight/feature/issue-42

Feature/issue 42 - Add routing code to work with doctor python types

Commit:
Added a few missing features from doctor v2.x

- Response validaiton
- Ability to specify different base handler on a route by route basis
- Ability to specify handler name on a route by route basis

Commit:
Merge branch 'feature/string-type-format' into feature/3.0

Commit:
Adding allowed_exceptions

Commit:
Add sphinx-autodoc-typehints

Commit:
Small amount of clean up

Commit:
Remove syslog specific check for response validation warning log

boralyl committed:
Merge pull request #47 from upsight/feature/issue-45

Feature/issue 45 Adds a few missing features from doctor 2.x

Commit:
Don't consider non SuperType annotations as request parameters.

boralyl committed:
Merge pull request #49 from upsight/feature/issue-48

Don't consider non SuperType annotations as request parameters.

Commit:
Renamed module router_v3 to routing

Commit:
Added function to allow decorators/middleware to add param annotations

boralyl committed:
Code cleanup

boralyl committed:
Adding a test for add_param_annotation

Commit:
Adding a few documentation specific features.

- Added an `example` property to all python types.  This will be used by
  the documentation when creating example values to show in the api
  docs.
- Added the ability to specify a custom `title` for each route's http
  method.  This will override the default values the documentation would
  use.
- Added the ability to specify a custom `heading` for each route.  This
  will group all routes (and their http methods) under this heading in
  the api docs.

boralyl committed:
Merge pull request #52 from upsight/feature/issue-46

Adding a few documentation specific features.

Commit:
Made docs generation work with python types.

Still need to clean up old code and tests.

Commit:
Fixing tests

Commit:
Removing old code and tests

Commit:
Renaming test file

Commit:
Removing old example app

Commit:
Renamed example app module

Commit:
Removing unused methods

Commit:
Updating documentation

Commit:
Added missing description/example to example Object type

Commit:
Updating intro

boralyl committed:
Merge pull request #53 from upsight/feature/make-docs-work

Feature/make docs work

Commit:
Merge branch 'feature/cleanup' into feature/3.0

Commit:
Merge branch 'feature/issue-43' into feature/3.0

Commit:
Removing six requirement

Commit:
Releasing 3.0.0-beta

Commit:
Updated readthedocs config file to specify python 3 support

Commit:
Releasing v3.0.0-beta2

noonat commented on CHANGELOG.rst:
```diff
@@ -4,6 +4,13 @@ Release History
+------------------------
+
+- First beta release of 3.0. This is a backwards incompatible change.  It drops
+  support for python 2 and the usage of json schemas.  See the documentation
```
It might be worth noting that JSONSchema is still supported, but just through a different means, and not at the top level.

boralyl replied:
I'll update it to mention json_schema_type.

noonat commented on README.rst:
```diff
@@ -18,51 +19,52 @@ doctor can easily be installed using pip:
+    # mytypes.py
+    from doctor.types import integer, boolean, Object
+
+    FooId = integer('The foo ID.', example=1)
```
Maybe use the `= types.integer(...)` style here, since it's the user's first exposure to this syntax?

noonat replied:
It might also be good to add a comment above this section and before the class below explaining why you have the different styles for the types. Something like:

```python
# doctor provides helper functions to easily define simple types.
FooId = integer('The foo ID.', example=1)
FetchBars = boolean('A flag that indicates if we should fetch bars', example=True)

# You can also inherit from type classes to create more complex types.
class Foo(Object):
    description = 'A Foo object'
    example = {'foo_id': 1}
    properties = {'foo_id': FooId}
    required = ['foo_id']
    additional_properties = False
```

boralyl replied:
Great suggestion

noonat commented on README.rst:
```diff
@@ -18,51 +19,52 @@ doctor can easily be installed using pip:
+    # foo.py
+    from mytypes import Foo, FooId, FetchBars
+
+    def get_foo(foo_id: FooId, fetch_bars: FetchBars=False) -> Foo:
```
I would also add a comment here:

```python
# Note the type annotations on this function definition. This tells Doctor how
# to parse and validate parameters for routes attached to this logic function.
def get_foo(foo_id: FooId, fetch_bars: FetchBars=False) -> Foo: 
```

boralyl replied:
Good call, makes it easier to follow.

noonat commented on docs/_logic_functions.rst:
```diff
@@ -3,14 +3,17 @@
 The logic function signature will be used to determine what the request
 parameters are for the route/HTTP method and to determine which are required.
+Each parameter must specify a `type hint <https://docs.python.org/3/library/typing.html>`_
+which is a sublcass of one of the :ref:`builtin doctor types<doctor-types>`.  
```
Typo: subclass

noonat commented on docs/_routes.rst:
```diff
@@ -2,47 +2,10 @@ Creating Routes
-for the request and response definitions, and a dict of routes.  Full 
-documentation on this topic can be found in the :ref:`router-docs` documentation.
+logic function.  We define our routes by instantiating a :class:`~doctor.routing.Route`.  A :class:`~doctor.routing.Route` requires 2 arguments.  The first is the
+actual uri.  The second is a tuple of allowed :class:`~doctor.routing.HTTPMethod` s for the uri: :func:`~doctor.routing.get`, :func:`~doctor.routing.post`, 
```
It would be more correct to say that the first is a regexp pattern used to check if the request's path matches the route, but I'm not sure if that would just confuse things more... Sinatra's docs call it a calls it a "URL-matching pattern". Flask's docs call it a "URL rule".

boralyl replied:
While `URL rule` is what flask mentions in the docs, I think your suggestion of `URL-matching pattern` is more friendly for the documentation.  I've used flask on multiple projects and didn't realize the docs referred to the route as a URL rule, but I know what URL-matching pattern means.

noonat commented on docs/types.rst:
```diff
@@ -0,0 +1,394 @@
+  datatype with a string representation. For example `date`, to represent an 
+  ISO 8601 formatted date string.  The following formats are supported:
+
+    * `date` - Will format the string as a `datetime.datetime` instance.  Expects
```
This says "will format the string" but I think it means "will parse the string"

noonat commented on docs/types.rst:
```diff
@@ -0,0 +1,394 @@
+      the format `'%Y-%m-%d'`
+    * `date-time` - Will format the string as a `datetime.datetime` instance.
+      Expects a valid ISO8601 string.  e.g. `'2018-02-21T16:09:02Z'`
+    * `email` - Does basic validation that the string is an emai by checking
```
Typo: `emai` should be `email`

noonat commented on docs/types.rst:
```diff
@@ -0,0 +1,394 @@
+Attributes
+##########
+
+* :attr:`~doctor.types.Object.additional_properties` - If `True` additional
```
I personally found this difficult to read. Maybe something like this:

```rst
* :attr:`~doctor.types.Object.additional_properties` - If `True`, additional
  properties (that is, ones not defined in
  :attr:`~doctor.types.Object.properties`) will be allowed.
```

boralyl replied:
Yeah, agreed, not worded super well

noonat commented on doctor/resource.py:
```diff
@@ -67,6 +61,35 @@ def get_annotation(cls, fn):
+        self.return_annotation = logic._doctor_signature.return_annotation
+        self.title = title
+        if title is None:
+            if http_method.upper() == 'GET':
```
This might be nicer as a map at the module level, something like:

```python
http_method_titles = {
    'GET': 'Retrieve',
    'POST': 'Create',
    'PUT': 'Update',
    'DELETE': 'Delete'
}

# then:
self.title = http_method_titles.get(http_method.upper())
```

boralyl replied:
+1

noonat reviewed (changes requested):
Just some minor comments, this looks really awesome! I'll also try out the new version in a test app this week.

Commit:
Add common-sense default example values for types.

Commit:
Use get_example()

Commit:
Documentation fixes and addressing a few PR comments.

Commit:
Adding back example values until #58 is merged

boralyl commented:
Addressed the comments here in https://github.com/upsight/doctor/pull/59 so they are easier to see.

Commit:
Fixed typos

Commit:
Fixed docstring

boralyl committed:
Merge pull request #59 from upsight/bugfix/pr-comments

Documentation fixes and addressing a few PR comments.

boralyl committed:
Merge pull request #58 from upsight/feature/issue-57

Add common-sense default example values for types.

Commit:
Removing the f strings as it's causing our documentation builds to fail.

This is because readthedocs only builds on up to python3.5 environments
and f strings were added in python3.6.

Commit:
Fixed issue when lines is a bytes instance.

boralyl committed:
Merge pull request #60 from upsight/feature/remove-f-strings

Feature/remove f strings

Commit:
Removed flask specific logic from routing.py

This will allow us to add other backends besides flask.

Commit:
Updating documentation for response validation.

Also added some docs on where example values come from when generating
api documentation.

boralyl committed:
Merge pull request #61 from upsight/feature/other-backends

Removed flask specific logic from routing.py

Commit:
Releasing v3.0.0-beta.3

Commit:
Added code to generate more logical handler names.

In the previous version if a handler name wasn't specified it would be a
generic Handler1 (where the number was incremented with each new
handler).  This sometimes made debugging things difficult, especially in
tests and debugging api docs generation when there were errors.

Commit:
Fixed typo

boralyl committed:
Merge pull request #62 from upsight/feature/handler-name

Added code to generate more logical handler names.

Commit:
Coerce doctor types to their native types before passing to logic
functions.

Commit:
Fixed some typos

Commit:
Made validation errors better to work iwth in 400 exceptions.

- returns all missing required fields in the error message
- Specifies which parameter the error corresponds to
- returns all errors for all params in the exception's errobj attr

boralyl committed:
Merge pull request #65 from upsight/bugfix/issue-64

Coerce doctor types to their native types before passing to logic functions

Commit:
Releasing v3.0.0-b.4

boralyl committed:
Fixed some typos and renamed errobj to errors.

boralyl committed:
Removed the param name prefix from the error message.

It only gets added if there is a single error.  Otherwise the user can
get the param/error combo from the errors attribute.

boralyl committed:
Fixed error message

boralyl committed:
Merge pull request #66 from upsight/bugfix/issue-63

Made validation errors better to work with 400 exceptions.

boralyl committed:
Fixed a few bugs.

- Fixed doctor attempting to document non doctor type params (#70)
- String with format of date now returns datetime.date (#69)
- Fixed swallowing of TypeError from SuperType class in Object __init__
  (#68)
- Changed the flask code to only raise response validation errors if an
  environment variable is set.  Before it also raised them when DEBUG
  was True in the config.  In practice this was incredibly annoying and
  slowed down development.  Especially in the case where a datetime
  string was returned that didn't include timezone information.  Updated
  the docs to reflect this too.

boralyl committed:
Fixed issue where we could have duplicate handler names which caused an
exception when adding routes to flask restful

Commit:
Updated the JsonSchema type to work properly.

Previously this type did not work with doctor since it required
instantiating an instance in order to obtain the example and description
values of the type.  Now the `json_schema_type` helper function will
load the schema and set those attributes on the generated class for you.

I didn't update the example app, but locally I tested out all the json
schema types and they properly get converted to the right python types.

Commit:
Removed debug

Commit:
Releasing v3.0.0-beta.5

Commit:
Added screenshot of example api docs to README

Commit:
Fixed a bug when a route didn't define a heading.

Commit:
Fixed README link

Commit:
Fixed tests

boralyl committed:
Merge pull request #72 from upsight/bugfix/jsonschema

Bugfix/jsonschema - Updated the JsonSchema type to work properly.

boralyl committed:
Merge pull request #71 from upsight/bugfix/issue-70

Fixed a few bugs.

Commit:
Fixed a few bugs with `json_schema_type`.

Commit:
Simplified the code in json_schema_type.

Also removed the resolve parameter and now it will resolve automatically
if it see `$ref` in the definition.

Commit:
Also recursively resolve refs.

Commit:
Removed elif branch for resolving ref.

It's handled by ResoucrceSchema.resolve.

Commit:
Added some comments to clarify the resolving of references

boralyl committed:
Merge pull request #74 from upsight/bugfix/issue-73

Fixed a few bugs with `json_schema_type`.

Commit:
Fixed bug in issue #75.

If add_param_annotations attempted to add a param that already existed
in the function signature it was trying to modify a ValueError would be
raised due to the duplicate parameter.  This change just skips adding
the duplicate parameter and logs a warning.

Commit:
Made prev_parameters a dict instead of a list for quicker lookups

boralyl committed:
Merge pull request #76 from upsight/bugfix/issue-75

Fixed bug in issue #75.

Commit:
Fixed a few bugs discovered during integration with an api.

- Updated handle_http to parse query and form parameters from strings to
  their expected type before we do validation on them.
- Fixed issue where if multiple decorators were used on a logic function
  and each one added param annotations the outer most decorator would
  erase any param annotations added from the previous decorator.

Commit:
Adds nullable attribute to all types.

Commit:
Updating documentation

Commit:
Updating version and changelog

boralyl committed:
Merge pull request #78 from upsight/bugfix/issue-77

Bugfix/issue 77 Bugfixes

Commit:
Fixes 2 small bugs.

- Updates parsing of query/form params to parse null values properly.
  #81
- Makes a copy of the logic function to preserve doctor attributes if
  the logic function is shared between routes.  #80

boralyl committed:
Merge pull request #82 from upsight/bugfix/issue-81

Fixes 2 small bugs.

Commit:
Moved copy_func to utils.

Also udpated the changelog for the latest beta release.

Commit:
Bumped version

boralyl committed:
Merge branch 'master' into feature/3.0

Commit:
Fixing flake8
